### PR TITLE
feat(datasource/docker): Add support for Google Application Default Credentials

### PR DESCRIPTION
**Pull request by kvanzuijlen**
## Changes
This PR adds support for Google Application Default Credentials, which also includes GKE Workload Identity. The benefit of using this is that you don't need a service account key (long-lived token) and is a well documented/best practice solution that doesn't need extra custom work on the end user side.

## Context
As said in the section above, this PR alleviates the need for a long lived token and instead introduces an option to configure a well documented solution.

## Documentation (please check one with an [x])

- [x] I have updated the documentation, or
- [ ] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [x] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [x] No unit tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

_I'll run these changes on a real repository shortly!_

### CODE REVIEW

File: lib/modules/datasource/docker/google.ts
Comment (viceice) [inline lib/modules/datasource/docker/google.ts:6]: `pkg.dev`: are you sure it can also be a docker registry? I only know `gcr.io` as registry

File: lib/modules/datasource/docker/google.ts
Comment (viceice) [inline lib/modules/datasource/docker/google.ts:26]: combine to one debug log message

File: lib/modules/datasource/docker/google.ts
Comment (kvanzuijlen) [reply lib/modules/datasource/docker/google.ts:6]: Thx for reviewing! Yes. `pkg.dev` is the domain for [Artifact Registry](https://cloud.google.com/artifact-registry). URLs for Artifact Registry generally look like this:
`<location>-docker.pkg.dev`
Location can be a single region (`us-central1`) or multi-region (`europe`).

File: lib/modules/datasource/docker/google.ts
Comment (kvanzuijlen) [reply lib/modules/datasource/docker/google.ts]: Done!

**Comment by kvanzuijlen** [comment]
@viceice the code coverage (of course) fails. I don't know how much sense it makes to test my current changes but I can look into writing a test for it if desired. Let me know what you would like me to do!

File: lib/modules/datasource/docker/google.ts
Comment (viceice) [reply lib/modules/datasource/docker/google.ts:6]: then make the regex more strict to only match the docker part

**Review by viceice — COMMENTED** [review]
please run on a real repo

**Comment by kvanzuijlen** [comment]
@viceice I ran it against a real repository and found out that I made a mistake. With the change I made the onboarding PR showed the expected PRs that it would create. Thus proofing (to me) that it works with Google Application Default Credentials, since I didn't set any authorization configs for Artifact Registry.

**Comment by kvanzuijlen** [comment]
@viceice Could you please re-review?

**Comment by rarkins** [comment]
Needs deconflicting

**Comment by kvanzuijlen** [comment]
@rarkins done

File: lib/modules/datasource/docker/common.ts
Comment (viceice) [inline lib/modules/datasource/docker/common.ts:119]:
```suggestion
      opts.headers = { authorization: `Basic ${auth}` };
```
auth should never be null or undefined here

File: lib/modules/datasource/docker/common.ts
Comment (viceice) [inline lib/modules/datasource/docker/common.ts:114]: check accesstoken before this

**Review by viceice — COMMENTED** [review]
still broken pnpm lockfile

**Comment by kvanzuijlen** [comment]
@viceice Sorry about that, the pnpm lock file rebroke after I fixed it in my earlier commit. It's working now again and I also applied your PR feedback (should've checked it after applying my fixes). Thanks for taking the time to review my PR!

**Review by viceice — DISMISSED** [review]

**Comment by viceice** [comment]
code coverage is very important.

please no commit hooks

**Comment by kvanzuijlen** [comment]
@viceice Thanks for your response. I added the unit tests. Please let me know if these are up to the standards. Locally it now reports 100% code coverage. (except for the branches, which were already lower than 100%).

Thanks for the time you put into reviewing this PR and for your patience with me :)

**Comment by kvanzuijlen** [comment]
@viceice I'm willing to look into adding the missing branch (else) tests and getting that to 100% as well. Right now it reports these files to have untested branches (at least locally):
image

Let me know if this is something you'd like to get to 100% and I'll create an issue and will start working on it if desired.

File: lib/modules/datasource/docker/index.spec.ts
Comment (viceice) [inline lib/modules/datasource/docker/index.spec.ts:42]: why the if?

File: lib/modules/datasource/docker/index.spec.ts
Comment (viceice) [inline lib/modules/datasource/docker/index.spec.ts:43]: we've a mocked test util to mock with types, search code

Comment (kvanzuijlen) [reply lib/modules/datasource/docker/index.spec.ts:43]: I'm not able to get it to work with `mocked`. The closest I came required me to mock all methods of GoogleAuth, which I don't think is desired.

File: lib/modules/datasource/docker/index.spec.ts
Comment (kvanzuijlen) [reply lib/modules/datasource/docker/index.spec.ts:42]: without the if I got errors saying it was undefined (before any of the tests using it ran)

File: lib/modules/datasource/docker/index.spec.ts
Comment (kvanzuijlen) [reply lib/modules/datasource/docker/index.spec.ts]: Updated the tests to make use of mocked, not sure if this is the desired solution.

File: lib/modules/datasource/docker/index.spec.ts
Comment (kvanzuijlen) [reply lib/modules/datasource/docker/index.spec.ts]: With the updated tests I no longer need this if statement

**Comment by kvanzuijlen** [comment]
@viceice is there something I need to change or can this feature be merged? Thank you for your time!

**Comment by rarkins** [comment]
@kvanzuijlen it's best that you don't merge from main any more because that resets the tests to require approval again

File: lib/modules/datasource/docker/index.spec.ts
Comment (viceice) [inline lib/modules/datasource/docker/index.spec.ts:44]:
```suggestion
```
mocks are reset globally

https://github.com/renovatebot/renovate/blob/eb634083144b63d3f22d8320cc44265ee366a1e4/jest.config.ts#L234

File: lib/modules/datasource/docker/index.spec.ts
Comment (viceice) [inline lib/modules/datasource/docker/index.spec.ts:517]:
```suggestion
```
a 200 response should never return that header

File: lib/modules/datasource/docker/index.spec.ts
Comment (viceice) [inline lib/modules/datasource/docker/index.spec.ts:495]: remove, see other comment

Comment (kvanzuijlen) [inline lib/modules/datasource/docker/index.spec.ts:495]:
```suggestion
```

**Review by viceice — DISMISSED** [review]

**Comment by kvanzuijlen** [comment]
@viceice I hadn't run prettier after the last PR suggestions and merges, sorry about that.

File: lib/modules/datasource/docker/index.spec.ts
Comment (viceice) [inline lib/modules/datasource/docker/index.spec.ts:493]:
```suggestion
        .reply(200)
```

File: lib/modules/datasource/docker/index.spec.ts
Comment (viceice) [inline lib/modules/datasource/docker/index.spec.ts:513]:
```suggestion
        .reply(200)
```

**Review by viceice — APPROVED** [review]

**Pull request merged**
